File: Medium/2130. Maximum Twin Sum of a Linked List/Solution.java
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null) return head;
        ListNode cur = head.next;
        head.next = null;
        while(cur != null) {
            ListNode go = cur.next;
            cur.next = head;
            head = cur;
            cur = go;
        }
        return head;
    }

    public int pairSum(ListNode head) {
        ListNode fast = head, slow = head;
        while(fast != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode startReversed = reverseList(slow);
        int ans = 0;
        while(startReversed != null) {
            ans = Math.max(ans, head.val + startReversed.val);
            head = head.next;
            startReversed = startReversed.next;
        }
        return ans;
    }
}

File: Medium/2130. Maximum Twin Sum of a Linked List/Solution.py
def reverseList(head):
    if not head or not head.next:
        return head
    cur = head.next
    head.next = None

    while cur is not None:
        go = cur.next
        cur.next = head
        head = cur
        cur = go

    return head
class Solution(object):
    def pairSum(self, head):
        fast = head
        slow = head
        while fast is not None:
            fast = fast.next.next
            slow = slow.next

        startReversed = reverseList(slow)
        ans = 0
        while startReversed is not None:
            ans = max(ans, head.val + startReversed.val)
            head = head.next
            startReversed = startReversed.next

        return ans